### ISSUE TITLE
blob storage support via SAS key

### ISSUE THREAD

**Issue by moadben** [Azure/azure-sdk-for-go#464]
Hello everyone,

I was thinking about implementing some sort of SAS key authentication for the blob storage library. Like maybe create an interface that the current Client satisfies and creating a separate, smaller Client object that sends api requests via SAS instead and have the blobstorageclient object be able to take both? I started working on something like that locally to use personally and was wondering if that would be something of interest for you guys.

Cheers,
Moad.

EDIT: If you guys are interested the only conflict I have right now is the whole generate a SAS key with a BlobStorageClient thing which you can't really do with a SAS Authenticated Client. Otherwise the new client works perfectly fine with a (pretty-much) intact codebase :)

**Comment by moadben** [comment]
So if you guys are interested the only conflict I have right now is the whole generate a SAS key with a BlobStorageClient thing which you can't really do with a SAS Authenticated Client.

**Comment by salameer** [comment]
Hi Moad, thanks for contacting us, assigning someone to take a look at this and hopefully get back so to you

**Comment by marstr** [comment]
For further conversation on this topic, please see the issue linked above.

**Comment by ppanyukov** [comment]
We actually hit a scenario today when we'd want to use SAS with this library. So yes, SAS support would be very nice indeed.

@marstr : I'm not sure there are issues linked above that you refer to?

**Comment by marstr** [comment]
Ah, it looks like the issue it was referring to has been deleted so my comment has been orphaned.

This issue resumes being the appropriate place for conversation. Sorry about the confusion, @ppanyukov!

**Comment by moadben** [comment]
Hey @ppanyukov,

I have a forked version of the azure-sdk with a fully functional SAS client implemented. However given that this was something I worked on 6 months ago I'm sure there would be a ton of issues attempting to integrate it with the current SDK.

I'd be more than happy to look into updating it though if this is something both you and the dev team would be interested in!

Thanks,
Moad.

**Comment by ppanyukov** [comment]
Yes definitely @moadben, this should be very useful. 

The use case for us was, if this helps:

- use SAS to control access to a specific container in a storage account;
- the user of this container needs to list blobs, create blobs etc;
- we did not need capabilities to generate SAS tokens in golang for our scenario;
- the users of the containers are both .NET and golang programs;
- we couldn't use this approach because golang storage SDK does not support SAS;
- it was an urgent production situation for us and we didn't have time to hack and test the SDK;
- lucky we solved our problem without SAS this time.

It wasn't a show stopper for us in this particular case, but I can well see it might be for someone in the future.
